### ISSUE TITLE
test_security_curve triggers stack smashing error when built with -fstack-protector

### ISSUE THREAD

**Issue by garlick** [zeromq/libzmq#1364]
When building an RPM for zeromq-4.0.5 on a RHEL 6 system, our builder threw this error during `make check`:

```
DEBUG: *** stack smashing detected ***:
   /builddir/build/BUILD/zeromq-4.0.5/tests/.libs/lt-test_security_curve terminated
DEBUG: ======= Backtrace: =========
DEBUG: /lib64/libc.so.6(__fortify_fail+0x37)[0x7fb1db992817]
DEBUG: /lib64/libc.so.6(__fortify_fail+0x0)[0x7fb1db9927e0]
DEBUG: /builddir/build/BUILD/zeromq-4.0.5/tests/.libs/lt-test_security_curve[0x402309]
DEBUG: /lib64/libc.so.6(__libc_start_main+0xfd)[0x7fb1db8aed1d]
DEBUG: /builddir/build/BUILD/zeromq-4.0.5/tests/.libs/lt-test_security_curve[0x400e99]
DEBUG: ======= Memory map: ========
...
```

I can reproduce this if I build zeromq with `./configure CXXFLAGS="-fstack-protector"`, presumably what some RHEL 6 RPM macro does on our builder.

I found changing the `return (0)` at the end of test program to `exit (0)` seems to solve this.  That's as far as I took it.

**Comment by twhittock** [comment]
Possibly related: CZMQ self-test fails semi-regularly on travis CI, eg https://travis-ci.org/zeromq/czmq/jobs/55571247:

```
lt-test_security_curve: testutil.hpp:143: void expect_bounce_fail(void*, void*): Assertion `rc == 32' failed.
FAIL: test_security_curve
```

**Comment by filefrog** [comment]
I also ran into this problem while building RPMs on CentOS 6, but after some further research, it seems the stack smashing is in fact legitimate.  Changing the `return' in main() to an exit() call bypasses the canary check that -fstack-protector inserts into the code tree, so the problem stops being reported.

The root cause seems to be that the public / private key buffers are declared as char[40] when they should in fact be char[41].  I checked latest HEAD, and the test is correct there and runs fine under SSP.

For my immediate packaging needs, I applied the following patch via the RPM spec file: https://gist.github.com/filefrog/a4b86edf7dde75dde8ce
